scalardb/src/scalardb/elle_append_2pc.clj

```diff
@@ -0,0 +1,149 @@
+(ns scalardb.elle-append-2pc
+  (:require [clojure.string :as str]
+            [jepsen.checker :as checker]
+            [jepsen.client :as client]
+            [jepsen.generator :as gen]
+            [jepsen.tests.cycle.append :as append]
+            [cassandra.conductors :as cond]
+            [scalardb.core :as scalar])
+  (:import (com.scalar.db.api Get
+                              Put)
+           (com.scalar.db.io IntValue
+                             TextValue
+                             Key)
+           (com.scalar.db.exception.transaction
+            UnknownTransactionStatusException)))
+
+(def ^:private ^:const KEYSPACE "jepsen")
+(def ^:private ^:const TABLE "txn")
+(def ^:private ^:const DEFAULT_TABLE_COUNT 3)
+(def ^:private ^:const SCHEMA {:id                     :int
+                               :val                    :text
+                               :tx_id                  :text
+                               :tx_version             :int
+                               :tx_state               :int
+                               :tx_prepared_at         :bigint
+                               :tx_committed_at        :bigint
+                               :before_val             :text
+                               :before_tx_id           :text
+                               :before_tx_version      :int
+                               :before_tx_state        :int
+                               :before_tx_prepared_at  :bigint
+                               :before_tx_committed_at :bigint
+                               :primary-key [:id]})
+(def ^:private ^:const ID "id")
+(def ^:private ^:const VALUE "val")
+
+(defn- prepare-get
+  [table id]
+  (-> (Key. [(IntValue. ID id)])
+      (Get.)
+      (.forNamespace KEYSPACE)
+      (.forTable table)))
+
+(defn- prepare-put
+  [table id value]
+  (-> (Key. [(IntValue. ID id)])
+      (Put.)
+      (.forNamespace KEYSPACE)
+      (.forTable table)
+      (.withValue (TextValue. VALUE value))))
+
+(defn- get-value
+  [r]
+  (some-> r .get (.getValue VALUE) .get .getString .get))
+
+(defn- tx-insert
+  [tx table id value]
+  (.put tx (prepare-put table id value)))
+
+(defn- tx-update
+  [tx table id prev value]
+  (.put tx (prepare-put table id (str prev "," value))))
+
+(defn- tx-execute
+  [tx1 tx2 [f k v]]
+  (let [key_hash (hash k)
+        table (str TABLE (mod key_hash DEFAULT_TABLE_COUNT))
+        tx (if (= (mod key_hash 2) 0) tx1 tx2)
```

Comment:
Is the number of transactions always 2?
Wouldn't we use more than 2 transactions with 2PC tests?